spec/forms/application_form_spec.rb

```diff
@@ -54,4 +54,110 @@
+      expect(example_form.send("date_of_birth(3i)")).to eq(2)
+    end
+
+    it "aliases the (1i), (2i), and (3i) suffixed attribute accessors to year, month, and day, respectively" do
+      example_form.send("date_of_birth(1i)=", 2024)
+      example_form.send("date_of_birth(2i)=", 5)
+      example_form.send("date_of_birth(3i)=", 2)
+
+      expect(example_form.date_of_birth_year).to eq(2024)
+      expect(example_form.date_of_birth_month).to eq(5)
+      expect(example_form.date_of_birth_day).to eq(2)
+
+      example_form.send("date_of_birth_year=", 2025)
+      example_form.send("date_of_birth_month=", 6)
+      example_form.send("date_of_birth_day=", 3)
+
+      expect(example_form.send("date_of_birth(1i)")).to eq(2025)
+      expect(example_form.send("date_of_birth(2i)")).to eq(6)
+      expect(example_form.send("date_of_birth(3i)")).to eq(3)
+    end
+
+    it "adds getter and setter methods to set and read all fields as Date objects" do
+      example_form.date_of_birth_year = 2024
+      example_form.date_of_birth_month = 5
+      example_form.date_of_birth_day = 2
+
+      expect(example_form.date_of_birth).to eq(Date.new(2024, 5, 2))
+
+      example_form.date_of_birth = Date.new(2025, 6, 3)
+
+      expect(example_form.date_of_birth_year).to eq(2025)
+      expect(example_form.date_of_birth_month).to eq(6)
+      expect(example_form.date_of_birth_day).to eq(3)
+    end
+
+    describe "date attribute setter" do
+      it "casts strings into dates" do
+        example_form.date_of_birth = "2024-05-02"
+
+        expect(example_form.date_of_birth).to eq(Date.new(2024, 5, 2))
+      end
+
+      it "casts dates into dates" do
+        example_form.date_of_birth = Date.new(2024, 5, 2)
+
+        expect(example_form.date_of_birth).to eq(Date.new(2024, 5, 2))
+      end
+
+      it "casts times into dates" do
+        example_form.date_of_birth = Time.new(2024, 5, 2, 12, 0, 0, 0)
+
+        expect(example_form.date_of_birth).to eq(Date.new(2024, 5, 2))
+      end
+
+      it "doesn't cast other values" do
+        expect { example_form.date_of_birth = 123 }.to raise_error(ArgumentError)
+        expect { example_form.date_of_birth = 123.4 }.to raise_error(ArgumentError)
+        expect { example_form.date_of_birth = true }.to raise_error(ArgumentError)
+        expect { example_form.date_of_birth = false }.to raise_error(ArgumentError)
+      end
+    end
+
+    describe "date attribute getter" do
+      it "composes the individual date components into a date object" do
+        example_form.date_of_birth_year = 2024
+        example_form.date_of_birth_month = 5
+        example_form.date_of_birth_day = 2
+
+        expect(example_form.date_of_birth).to eq(Date.new(2024, 5, 2))
+      end
+
+      it "returns nil if any of the date components are missing" do
+        example_form.date_of_birth_year = 2024
+        example_form.date_of_birth_month = 5
+
+        expect(example_form.date_of_birth).to be_nil
+      end
```

Comment:
We have a slightly different implementation in placements, which came about as a result of a content review done on the service.

A content designer fed back that partially entered dates – for example, if a user filled in the "day" and "month" fields, but left the "year" field blank – resulted in all three fields being reset following validation. That was happening because invalid dates were returned as `nil` for the entire attribute – i.e. exactly as this code does.

The content designer considered this behaviour to be a bug. Instead, the desired behaviour was for populated fields to be persisted, even if others were left blank.

For more details, see these artefacts:

- Trello: [Add mentor → preserve partially entered DOB fields](https://trello.com/c/c7hMIOEu/486-add-mentor-%E2%86%92-preserve-partially-entered-dob-fields)
- https://github.com/DFE-Digital/itt-mentor-services/pull/761

FWIW I believe this 'buggy' behaviour just wasn't noticed on the Claims side, but I think it's fair to expect consistent behaviour between the two services here.

Let's assume this behaviour is fine for now – after all, that's the existing behaviour in Claims, and I don't think it'd be in scope to change as part of this PR. However I think it would be good to follow up and standardise all our existing date attribute behaviour to use this concern, at which point the behaviour would need to change accordingly.

Reply:
I've addressed this in the latest fixup.

![CleanShot 2024-08-23 at 11 37 14](https://github.com/user-attachments/assets/3328021d-480e-48a9-a15f-73d0beb4fef8)

Practically speaking, it's a bit strange, which makes the implementation also strange.

Ideally, we'd have a partial-`Date`-like class that interacts with other `Date` instances.

This would allow us to avoid this code:

```ruby
if year.present? && month.present? && day.present?
  Date.new(year.to_i, month.to_i, day.to_i)
else
  IncompleteDate.new(year, month, day)
end
```

and just return...

```ruby
PotentiallyIncompleteDate.new(year, month, day)
```

`PotentiallyIncompleteDate` would behave exactly like a `Date` when it's components are complete but would respond to comparisons with other `Date` or `Date`-like objects, but return `false`.

`Date.new(2024, 5, 1) < PotentiallyIncompleteDate.new(2024, nil, 1) == false # always`

This would remove the need for this attribute method:

```ruby
define_method "#{attribute_name}_is_date?" do
  send(attribute_name).is_a?(Date)
end
```

and the usage of this method on all date-related validations (that are usually skipped when they're `nil`):

```ruby
validates :ends_on, comparison: { greater_than_or_equal_to: :starts_on }, if: -> { starts_on_is_date? && ends_on_is_date? }

validate :does_not_start_within_another_claim_window, if: :starts_on_is_date?
validate :does_not_end_within_another_claim_window, if: :ends_on_is_date?
```

Unfortunately, `#present?` returning `false` would be weird, so a `date_presence` validator, or similar, would be required to show "presence" errors without being blank.

```ruby
validates :starts_on, date_presence: true
validates :ends_on, date_presence: true
```

-----

To avoid investing too much time into creating a `Date`-interoperable class, I've added the attribute method and the conditional validations.